### PR TITLE
[ADD] 14.0 hr_attendance_validation

### PR DESCRIPTION
**Pull request by petrus-v**
This add a  validation mechanism to review employee attendance and generate compensatory hours (allocated leaves) that can be used later or regulate credits as this module is  compatible with [hr_holidays_credit](https://github.com/OCA/hr-holidays/tree/14.0/hr_holidays_credit).

This PR is based on top of #50 allow manager to review attendance and generate compensatory hours attribution or leaves based on weekly overtime.

### CODE REVIEW

File: hr_attendance_validation/models/hr_attendance.py [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+                    )
+                    % (
+                        record.employee_id.name,
+                        record.check_in.date(),
+                    )
+                )
+        return res
+
+    def _is_validated_employee_week(self):
+        validated_week = self.env["hr.attendance.validation.sheet"].search_count(
```
Comment (petrus-v) [inline hr_attendance_validation/models/hr_attendance.py]: this search can occur with users that are not allowed to read on that model, manage access or use super user to perform this search count.

**Comment by petrus-v** [comment]
@OCA/human-resources-maintainers I've just rebase but it would require someone to explicitly mark as no stale if possible :pray:  ? and probably add the `needs reviews` tags ;)

Thanks deer maintainers !

**Comment by petrus-v** [comment]
just rebase this PR as #50 has been merged

File: hr_attendance_validation/models/hr_attendance.py [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+
+
+class HrAttendance(models.Model):
+    _inherit = "hr.attendance"
+
+    is_overtime_due = fields.Boolean(
+        string="Is overtime due",
+        default=False,
+        help="What ever the overtime is due or not. "
+        "By default overtime are not due until manager validate them.",
```
Comment (HekkiMelody) [inline hr_attendance_validation/models/hr_attendance.py]:
```suggestion
        help="Whether the overtime is due or not. "
        "By default overtime is not due until a manager validates it.",
```
There may be a better word than "due" to make the concept clearer, but I can't think of it...

File: hr_attendance_validation/models/hr_attendance.py [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+    @api.model
+    def create(self, *args, **kwargs):
+        attendance = super().create(*args, **kwargs)
+        if attendance._is_validated_employee_week():
+            raise ValidationError(
+                _(
+                    "Can not create new attendance for employee %s. "
+                    "Attendance validation week has been already reviewed and validated "
+                    "for the given check_in date %s"
```
Comment (HekkiMelody) [inline hr_attendance_validation/models/hr_attendance.py]:
```suggestion
                    "Cannot create new attendance for employee %s. "
                    "Attendance for the day of the check in %s "
                    "has already been reviewed and validated."
```

File: hr_attendance_validation/models/hr_attendance.py [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+                        record.check_in.date(),
+                    )
+                )
+        res = super().write(*args, **kwargs)
+        for record in self:
+            if record._is_validated_employee_week():
+                raise ValidationError(
+                    _(
+                        "Can not change this attendance (%s, %s) "
+                        "which would be moved to a validated week."
```
Comment (HekkiMelody) [inline hr_attendance_validation/models/hr_attendance.py]:
```suggestion
                        "which would be moved to a validated day."
```

File: hr_attendance_validation/models/hr_attendance_validation_sheet.py [hunk 5]
```diff
@@ -0,0 +1,345 @@
...
+from datetime import timedelta
+
+from odoo import SUPERUSER_ID, _, api, fields, models
+from odoo.osv import expression
+
+
+class HrAttendanceValidationSheet(models.Model):
+    _name = "hr.attendance.validation.sheet"
+    _description = (
+        "Attendance validation sheet that helps " "manager to review attendances."
```
Comment (HekkiMelody) [inline hr_attendance_validation/models/hr_attendance_validation_sheet.py]:
```suggestion
        "Attendance validation sheet that helps managers review attendances."
```

File: hr_attendance_validation/models/hr_attendance_validation_sheet.py [hunk 6]
```diff
@@ -0,0 +1,345 @@
...
+    )
+    _order = "date_from desc, employee_id asc"
+
+    def _default_from_date(self):
+        """return the last monday before the last sunday"""
```
Comment (HekkiMelody) [inline hr_attendance_validation/models/hr_attendance_validation_sheet.py]:
```suggestion
        """returns the monday before last sunday"""
```

File: hr_attendance_validation/models/hr_attendance_validation_sheet.py [hunk 7]
```diff
@@ -0,0 +1,345 @@
...
+        today = fields.Date.today()
+        return today - timedelta(days=today.weekday() + 7)
+
+    def _default_to_date(self):
+        """return last passed sunday"""
```
Comment (HekkiMelody) [inline hr_attendance_validation/models/hr_attendance_validation_sheet.py]:
```suggestion
        """returns last sunday"""
```

File: hr_attendance_validation/models/hr_attendance_validation_sheet.py [hunk 8]
```diff
@@ -0,0 +1,345 @@
...
+        help="Compensatory hour that will be allocated if > 0 to the employee.",
+    )
+    regularization_compensatory_hour_taken = fields.Float(
+        "Regularization compensatory hours' taken",
+        help="Compensatory hours that will be apply as leaves if > 0 to the employee.",
+    )
+    require_regeneration = fields.Boolean(
+        "Require regeneration",
+        default=False,
+        help="Couldn't properly call action retreive lines in onchange "
```
Comment (HekkiMelody) [inline hr_attendance_validation/models/hr_attendance_validation_sheet.py]:
```suggestion
        help="Couldn't properly call action retrieve lines in onchange "
```

File: hr_attendance_validation/models/hr_attendance_validation_sheet.py [hunk 9]
```diff
@@ -0,0 +1,345 @@
...
+        compute="_compute_attendances_hours",
+        help="Compute number of attendance lines marked as overtime which are not due",
+    )
+    compensatory_hour = fields.Float(
+        "Compensatory hour",
+        help="Compensatory hour that will be allocated if > 0 to the employee.",
+    )
+    regularization_compensatory_hour_taken = fields.Float(
+        "Regularization compensatory hours' taken",
+        help="Compensatory hours that will be apply as leaves if > 0 to the employee.",
```
Comment (HekkiMelody) [inline hr_attendance_validation/models/hr_attendance_validation_sheet.py]:
```suggestion
        help="Compensatory hours that will be counted as leaves for the employee.",
```

File: hr_attendance_validation/models/hr_attendance_validation_sheet.py [hunk 10]
```diff
@@ -0,0 +1,345 @@
...
+        help="Compute number of attendance lines marked as overtime which are marked as due",
+    )
+    overtime_not_due_hours = fields.Float(
+        "Overtime not due (hours)",
+        compute="_compute_attendances_hours",
+        help="Compute number of attendance lines marked as overtime which are not due",
+    )
+    compensatory_hour = fields.Float(
+        "Compensatory hour",
+        help="Compensatory hour that will be allocated if > 0 to the employee.",
```
Comment (HekkiMelody) [inline hr_attendance_validation/models/hr_attendance_validation_sheet.py]:
```suggestion
        help="Compensatory hours that will be allocated to the employee.",
```

File: hr_attendance_validation/views/res_config_settings_views.xml [hunk 11]
```diff
@@ -0,0 +1,42 @@
...
+                        <div
+                            class="col-12 col-lg-6 o_setting_box"
+                            title="Set compensatory leave type"
+                        >
+                            <div class="o_setting_left_pane">
+                            </div>
+                            <div class="o_setting_right_pane">
+                                <span class="o_form_label">Leave type</span>
+                                <div class="text-muted">
+                                    This is used while allocate new leaves from attendance sheet review validaiton.
```
Comment (HekkiMelody) [inline hr_attendance_validation/views/res_config_settings_views.xml]:
```suggestion
                                    This view is used to allocate new leaves during the process of review and validation of the attendance sheet.
```

**Review by HekkiMelody — APPROVED** [review]
Just a little cleanup here and there.
Could you also check that all the instances of `retreive` become `retrieve`?

Aside from that, looks like a very useful and well-made module. Thanks so much for your work!

File: hr_attendance_validation/models/hr_attendance.py (hunk 2 above)
Comment (petrus-v) [reply hr_attendance_validation/models/hr_attendance.py]: does "owed" is proper than "due" ?

File: hr_attendance_validation/views/res_config_settings_views.xml (hunk 11 above)
Comment (petrus-v) [reply hr_attendance_validation/views/res_config_settings_views.xml]:
```suggestion
                                    This leave type is used to allocate new leaves during the process of review and validation of the attendance sheet.
```

**Comment by petrus-v** [comment]
thanks a lot for review and suggestions @aleuffre I'll check `retreive` instance and update translations according your suggestions

**Comment by HekkiMelody** [comment]
Thank you for your work! @petrus-v  

This module seems very useful for my use case. I gave it a spin on runboat and didn't find any obvious problems, but I'm going to try it more thoroughly in the following days and weeks as I work to integrate it into an odoo instance and will report if I find anything.

**Comment by petrus-v** [comment]
I've squashed commit and rebase on top of 14.0 branch.

**Comment by dreispt** [comment]
@OCA/human-resources-maintainers is this good to merge?

**Comment by nimarosa** [comment]
Hello, code LGTM. If anyone can make functional tests I can merge it.

**Review by nimarosa — APPROVED** [review]

**Comment by nimarosa** [comment]
I saw there was a functional approval in earlier comments. We can merge. 

/ocabot merge nobump

**Pull request merged**